rgw multisite: fixes for meta sync across periods

cbodley opened the pull request:
fixes for the incremental metadata sync logic that detects the end of one period and advances to the next

the `rgw: RGWMetaSyncShardCR uses fake error to break out of RGWBackoffControlCR` commit is admittedly hacky, but i couldn't find a good way to modify the `RGWBackoffControlCR` interface to support this use (should `RGWBackoffControlCR` really keep retrying on success? suggestions welcome!)

passes the updated `test_multi_period_incremental_sync()` test in https://github.com/ceph/ceph/pull/13067

cbodley commented:
added http://tracker.ceph.com/issues/18639 for backport

yehudasa commented on src/rgw/rgw_period_history.cc:
```diff
@@ -69,6 +69,15 @@ bool Cursor::has_next() const
+
+bool operator!=(const Cursor& lhs, const Cursor& rhs)
+{
+  return lhs.history != rhs.history || lhs.epoch != rhs.epoch;
```
or can just do `return !(lhs == rhs)`

cbodley commented:
[teuthology results](http://pulpito.ceph.com/cbodley-2017-01-26_14:41:45-rgw-wip-cbodley-testing---basic-mira/): all green except for 10 dead jobs

cbodley commented:
jenkins test this please

cbodley commented:
@yehudasa look okay to merge?

yehudasa commented:
@cbodley there are some cases that I'm not sure we really cover, and maybe we need to  think about in the context of this PR: when we switch to a new period, the max marker of the previous period should be the point at which the new master was, and not necessarily where the original master was. I don't think we currently do a rollback of changes that happened on original master, but lost due to the switch to the new master. This can affect all the zones, and not necessarily the new master.

cbodley commented on src/rgw/rgw_sync.cc:
```diff
@@ -1576,10 +1571,6 @@ class RGWMetaSyncShardCR : public RGWCoroutine {
                                                &max_marker, INCREMENTAL_MAX_ENTRIES,
                                                &log_entries, &truncated));
           for (log_iter = log_entries.begin(); log_iter != log_entries.end(); ++log_iter) {
-            if (!period_marker.empty() && period_marker < log_iter->id) {
```
it looks like my diagnosis in this commit was wrong - this line is not comparing markers from different periods. rather, if the old and new master zones agree on the last marker of this period, we won't see any log entries with a marker greater than `period_marker`, and this block will never execute

cbodley commented:
@yehudasa updated to fix handling of the period markers - test_multi.py is passing reliably (including the updated test in https://github.com/ceph/ceph/pull/13067)

i also changed the interaction between `RGWBackoffControlCR` and `RGWMetaSyncShardCR`. `RGWBackoffControlCR` no longer retries on success, so `RGWMetaSyncShardCR` can just return success when it reaches the end of a period. this breaks out of the `RGWBackoffControlCR` loop and returns control back to `RGWMetaSyncCR`, which can advance to the next period. the previous version of this pull request was returning a fake error in this case, but https://github.com/ceph/ceph/issues/13546 recently merged that causes it to retry on all errors

this change to `RGWBackoffControlCR` exposed some odd behavior in `RGWDataSyncCR` that relied on retrying after success. in the first call to `RGWDataSyncCR`, `RGWInitDataSyncStatusCoroutine` would create the sync status objects and return success. but it didn't initialize the `sync_status.sync_markers` used to spawn `RGWDataSyncShardControlCR`, so we'd loop over an empty map, spawn nothing, and return success. once `RGWBackoffControlCR` called us back, we would re-read the sync status object to get the correct `sync_status.sync_markers`, and continue as normal.

i modified `RGWInitDataSyncStatusCoroutine` a bit so that `RGWDataSyncCR` could pass in a pointer to its `rgw_data_sync_status` to avoid having to read it back from rados

cbodley commented:
updates:
* `rgw_sync_status_marker` includes a `realm_epoch` so that markers from old periods can be ignored in comparisons against other periods
* `RGWPeriod::update_sync_status()` reads markers from its sync status using `RGWMetaSyncStatusManager::read_sync_status()`, rather than getting the latest markers from its local mdlog (which would include entries that it hasn't synced)

passes tests in https://github.com/ceph/ceph/pull/13067 (which have a new `sleep()` after period commit, which is expected to race with metadata sync by design)

cbodley commented:
rebased, ping @yehudasa

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -1427,33 +1427,48 @@ void RGWPeriod::fork()
+  const auto current_epoch = current_period.get_realm_epoch();
+  if (current_epoch != status.sync_info.realm_epoch) {
+    // no sync status markers for the current period
+    const int behind = current_epoch - status.sync_info.realm_epoch;
```
@cbodley is it possible somehow that current_epoch < status.sync_info.realm_epoch?

cbodley replied:
probably not possible, because `RGWPeriod::commit()` has a check for `if (realm_epoch != current_period.get_realm_epoch() + 1)` just prior to calling `update_sync_status()`. that would mean that meta sync had advanced past the current_period

yehudasa replied:
@cbodley how about adding assert there to cover that case?

cbodley replied:
added `assert(current_epoch > status.sync_info.realm_epoch);`

cbodley commented:
jenkins test this please

cbodley commented:
rebased and still passing `test_multi_period_incremental_sync()`

the only conflicts were in `test_multi.py`, so i'm not planning to run through teuthology again

cbodley commented:
@yehudasa @mattbenjamin could one of you please override the failed build check? the only `make check` failure was `148 - osd-scrub-repair.sh (Failed)`
